Rpw/initial

rpwils opened the pull request:
Team, This is the initial commit and is expected to have many iterations to build out as we discussed. Please take a look at the code, documentation and make recommendations. I have left some blank sections in the readme for us to discuss.  
the next task is to convert to use numpy arrays for all functions and update test to use a file for the data.

rpwils committed:
initial checkin

rpwils committed:
black formatted

rpwils committed:
docs: implemented numpydocs formatting

rpwils committed:
docs: updated readme

scaubrey commented on .travis.yml:
```diff
@@ -0,0 +1,10 @@
+language: python
+python:
+  - "3.7"
```
Will be putting this in the project template, but we decided we'll be supporting python 3.6 and 3.8 as well.

ed-nykaza replied:
+1, please add this to travis and the template.

scaubrey commented on tidepool_data_science_metrics/cgm/cgm.py:
```diff
@@ -0,0 +1,187 @@
+    int
+        The number of episodes matching input specifications.
+    """
+    bg_values_df.loc[(bg_values_df["values"] < episodes_threshold), "episode"] = 1
+    bg_values_df.loc[(bg_values_df["values"] >= episodes_threshold), "episode"] = 0
+    bg_values_df["group"] = 0
+    bg_values_df["group"][
+        (
+            (bg_values_df.episode == 1)
+            & (bg_values_df.episode.shift(1) == 0)
+            & (bg_values_df.episode.shift(-1) == 1)
+            & (bg_values_df.episode.shift(-(min_ct_per_ep - 1)) == 1)
+        )
+    ] = 1
+    bg_values_df["group"][
+        (
+            (bg_values_df.episode == 1)
+            & (bg_values_df.index == 0)
+            & (bg_values_df.episode.shift(-1) == 1)
+            & (bg_values_df.episode.shift(-2) == 1)
+        )
+    ] = 1
```
I have a general idea of what's going on here, but it'd be great to add plain language comments detailing each of these masks and calls. Updating the docstring would probably help too, but I know this is a work in progress

scaubrey commented on tidepool_data_science_metrics/common/common.py:
```diff
@@ -0,0 +1,60 @@
+    return round(np.mean(bg_values), round_val)
+
+
+def avg(bg_values, round_val=2):
+    """
+    Calculate the average within a set of glucose values
+
+    Parameters
+    ----------
+    bg_values : ndarray
+        1D array containing data with `int` type.
+    round_val : int
+        The number of digits to round the result to.
+
+    Returns
+    -------
+    int
+        The calculated Average
+    """
+
+    return round(np.average(bg_values), round_val)
```
I had to look up why numpy had both mean and avg, and according to SO it's because average optionally accepts weights. Perhaps we can do the same here?

ed-nykaza replied:
yeah, that would be great

scaubrey reviewed (commented):
Looking pretty good, made a few comments

rpwils committed:
converted episodes to numpy and started refactor UT data.

rpwils committed:
fix : fixed bug in episodes where middle values in a range were ignored.

ed-nykaza commented on README.md:
```diff
@@ -2,18 +2,12 @@
 
```
There is a better project template as of this morning: https://github.com/tidepool-org/data-science-project-template/blob/master/README.md

rpwils replied:
added

ed-nykaza commented on pypi_push.sh:
```diff
@@ -0,0 +1,9 @@
+##STOP - Make sure you updated the version in the setup.py to match the version in github to be uploaded.
```
is it typically for the versioning to be a manual process, or are there automated tools? I am not suggesting that we need to do this now, just curious?

ed-nykaza commented on requirements.txt:
```diff
@@ -1,8 +1,11 @@
-matplotlib==3.2.1
```
is there a reason why flake8 is not included?

rpwils replied:
added

ed-nykaza commented on setup.py:
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
```
is this old and outdated?

ed-nykaza commented on tests/cgm/test_metrics.py:
```diff
@@ -0,0 +1,167 @@
+
+
+def get_date_values():
+    values = [
```
doesn't have to be done now, but I would prefer to see this data loaded in from a csv or created on the fly.

rpwils replied:
yes will do that in a followup change

rpwils replied:
updated to csv

ed-nykaza commented on tests/cgm/test_metrics.py:
```diff
@@ -0,0 +1,167 @@
+def get_date_ep_values():
+    return np.array(
+        [
+            ["8/15/2019 00:40:00", 87],
```
same comment as above (load from csv)

ed-nykaza commented on tests/conftest.py:
```diff
@@ -0,0 +1,110 @@
+def bg_values(scope="module"):
+    return np.array(
+        [
+            [100],
```
same as above

ed-nykaza commented on tidepool_data_science_metrics/__init__.py:
```diff
@@ -0,0 +1,19 @@
+from tidepool_data_science_metrics.common import common
+
+"""   
+from tidepool_data_science_metrics.cgm2 import (
```
where does cgm2 come from?

ed-nykaza commented on tidepool_data_science_metrics/__init__.py:
```diff
@@ -0,0 +1,19 @@
+
+"""   
+from tidepool_data_science_metrics.cgm2 import (
+    cgm3.get_percent_values_by_range,
```
what is cgm3

ed-nykaza reviewed (changes requested):
@rpwils I will do another deeper review on the actual functions (soon). This review was just a quick syntax/process review.

ed-nykaza commented on tidepool_data_science_metrics/cgm/cgm.py:
```diff
@@ -0,0 +1,179 @@
+import tidepool_data_science_metrics.common.common as common
+
+
+def cv_of_glucose(bg_values, round_val=2):
```
this can go in commons with the name `coefficient_of_variation`

ed-nykaza commented on tidepool_data_science_metrics/cgm/cgm.py:
```diff
@@ -0,0 +1,179 @@
+
+
+def gmi(bg_values, round_val=2):
+    """
```
it would be ideal to have citations to all of these functions.

rpwils replied:
done

ed-nykaza commented on tidepool_data_science_metrics/cgm/cgm.py:
```diff
@@ -0,0 +1,179 @@
+
+
+def percent_values_by_range(
+    bg_values, lower_threshold: int, upper_threshold: int, round_val=2
```
if we renamed bg_values to bg_array it might be easier for people to know what they need to pass in

rpwils replied:
updated

ed-nykaza commented on tidepool_data_science_metrics/cgm/cgm.py:
```diff
@@ -0,0 +1,179 @@
+    return round(gmi, round_val)
+
+
+def percent_values_by_range(
```
I like the generalization of percent_values_by_range, and also think we should add in specific and common ranges (e.g., percent_values_in_range_70-180, percent_values_below_<>, where <> can be 40, 54, 70. 

A great test to add would be to calculate % values < 70, between 70-180, and > 180 and then to make sure that the total of all 3 percentages added up to 100%.

ed-nykaza replied:
here is a bunch that we can add immediately % Time in target range (70 – 180 mg/dL)
% Time > 180 mg/dL
Mean glucose
% Time < 70 mg/dL
% Time < 54 mg/dL
% Time > 250

rpwils replied:
when you say specific and common ranges...do you mean have it auto generate those or have different functions for each ?

rpwils replied:
added the functions

rpwils committed:
fix : removed percent time in range as it is not calculating over time.

rpwils committed:
feat : updated many small recommendations from team including adding weights to numpy average.

rpwils committed:
feat : updated many small recommendations from team including adding weights to numpy average.

rpwils committed:
feat : added the following functions
percent_time_in_range_70_180
percent_time_above_180
percent_time_below_70
percent_time_below_54
percent_time_above_250

rpwils committed:
refactor : moved cv_of_glucose to common and renamed to coefficient_of_variation. Also added a citation

rpwils committed:
fix : working on fixing travis build

rpwils committed:
fix : working on fixing travis build

rpwils committed:
fix : file path for test

rpwils committed:
chore : updated release for test

rpwils committed:
feat : added warning for high and low bg values that are unreasonable.

rpwils committed:
refactor : refactored bgri function name and change the 2d array to 1d for episodes

rpwils committed:
fix : corrected the percent time above 250

rpwils committed:
refactor : multiple changes - Tox, Types, removed dead code.

ed-nykaza commented on requirements.txt:
```diff
@@ -1,4 +1,6 @@
 black==19.10b0
+flake8==3.7.9
+mypy==0.770
```
are we using mypy? if not, we should remove this, right?

rpwils replied:
yes we have mypy in it

ed-nykaza commented on tidepool_data_science_metrics/cgm/cgm.py:
```diff
@@ -1,104 +1,196 @@
+        The percent values above 180.
+    """
+    return percent_values_by_range(
+        bg_array, lower_threshold=180, upper_threshold=1000, round_val=round_val
```
this needs to be 180.5 instead of 180, as it is the percent of values > 180.

rpwils replied:
going to use >

ed-nykaza commented on tidepool_data_science_metrics/cgm/cgm.py:
```diff
@@ -1,104 +1,196 @@
+        The percent values below 70.
+    """
+    return percent_values_by_range(
+        bg_array, lower_threshold=0, upper_threshold=70, round_val=round_val
```
69.499 instead of 70, and I think the lower threshold needs to be 1 instead of 0.

ed-nykaza commented on tidepool_data_science_metrics/cgm/cgm.py:
```diff
@@ -1,104 +1,196 @@
+    """
+    _validate_bg(bg_array)
+    return percent_values_by_range(
+        bg_array, lower_threshold=0, upper_threshold=54, round_val=round_val
```
53.499

ed-nykaza replied:
and 1

rpwils replied:
updated

ed-nykaza commented on tidepool_data_science_metrics/cgm/cgm.py:
```diff
@@ -1,104 +1,196 @@
+        The percent values above 250.
+    """
+    return percent_values_by_range(
+        bg_array, lower_threshold=250, upper_threshold=1000, round_val=round_val
```
250.5

rpwils replied:
going to use <

ed-nykaza commented on tests/cgm/test_cgm.py:
```diff
@@ -0,0 +1,207 @@
+import pytest
```
please add in the test where you take the bg_array and caluclate the percent < 70, percent >=70 and <= 180, and the percent > 180. The test should sum those three percentages and should exactly equal 100%.

rpwils replied:
added test

ed-nykaza reviewed (changes requested):
@rpwils in general, it looks good. Please see the minor comments and the one major comment re: make sure that < 70, 54, and 40 are less than and not less than or equal to and same for > 180, 250, 300, and 400. Those should be greater than those values and not greater than or equal. Lastly, please change round_val to round_to_ndigits, and please run black and flake8 on the files. Please let me if you have questions. -ed

rpwils committed:
refactor : multiple changes - names changes, percent in time uses < and >.

rpwils committed:
updated version
